File: api/src/main/java/ecurrencies/api/EcurrencyServiceException.java
package ecurrencies.api;

public class EcurrencyServiceException extends Exception {

	private static final long serialVersionUID = 2323133755277870948L;

	private final boolean recoverable;

	public EcurrencyServiceException(boolean recoverable) {
		super();
		this.recoverable = recoverable;
	}

	public EcurrencyServiceException(boolean recoverable, String message) {
		super(message);
		this.recoverable = recoverable;
	}

	public EcurrencyServiceException(boolean recoverable, Throwable cause) {
		super(cause);
		this.recoverable = recoverable;
	}

	public EcurrencyServiceException(boolean recoverable, String message,
			Throwable cause) {
		super(message, cause);
		this.recoverable = recoverable;
	}

	public EcurrencyServiceException(boolean recoverable, String message,
			Throwable cause, boolean enableSuppression,
			boolean writableStackTrace) {
		super(message, cause, enableSuppression, writableStackTrace);
		this.recoverable = recoverable;
	}

	public boolean isRecoverable() {
		return recoverable;
	}

}

File: api/src/main/scala/ecurrencies/api/EcurrencyServiceException.scala
package ecurrencies.api

@SerialVersionUID(1L)
class EcurrencyServiceException(recoverable: Boolean, cause: Throwable) extends Exception(cause) with Serializable {

  def this(recoverable: Boolean) = this(recoverable, null)

  lazy val isRecoverable = recoverable
}

object EcurrencyServiceException {
  def apply(recoverable: Boolean, cause: Throwable) = new EcurrencyServiceException(recoverable, cause)

  def apply(recoverable: Boolean) = new EcurrencyServiceException(recoverable)
}